packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+        chainId,
+        address(this)
+      )
+    );
+    emit EIP712DomainSeparatorSet(eip712DomainSeparator);
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Returns info about attestations (with unrevoked signers)
+   *   for an identifier produced by a list of issuers
+   * @param identifier Hash of the identifier
+   * @param trustedIssuers Array of n issuers whose attestations will be included
+   * @param maxAttestations Limit the number of attestations that will be returned
+   * @return for m found attestations, m <= maxAttestations:
+   *         [
+   *           Array of m accounts,
+   *           Array of m issuedOns,
+   *           Array of m signers
+   *         ]; index corresponds to the same attestation
+   * @dev Adds attestation info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of attestations per issuer
+  function lookupAttestations(
+    bytes32 identifier,
+    address[] memory trustedIssuers,
+    uint256 maxAttestations
+  ) public view returns (address[] memory, uint256[] memory, address[] memory) {
+    // Cannot dynamically allocate an in-memory array
+    // For now require a max returned parameter to pre-allocate and then shrink
+    // TODO ASv2 is it a risk to allocate an array of size maxAttestations?
+    // Same index corresponds to same attestation
+    address[] memory accounts = new address[](maxAttestations);
+    uint256[] memory issuedOns = new uint256[](maxAttestations);
+    address[] memory signers = new address[](maxAttestations);
+
+    uint256 currIndex = 0;
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      uint256 numTrustedIssuers = identifierToAddresses[identifier][trustedIssuers[i]].length;
+      for (uint256 j = 0; j < numTrustedIssuers; j = j.add(1)) {
+        // Only create and push new attestation if we haven't hit max
+        if (currIndex < maxAttestations) {
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation memory attestation = identifierToAddresses[identifier][trustedIssuers[i]][j];
+          if (!revokedSigners[attestation.signer]) {
+            accounts[currIndex] = attestation.account;
+            issuedOns[currIndex] = attestation.issuedOn;
+            signers[currIndex] = attestation.signer;
+            currIndex = currIndex.add(1);
+          }
+        } else {
+          break;
+        }
+      }
+    }
+
+    // Trim returned structs if necessary
+    if (currIndex < maxAttestations) {
+      address[] memory trimmedAccounts = new address[](currIndex);
```

Comment:
This is the motivation for the updates to the lookup functions ([this ticket](https://app.zenhub.com/workspaces/identity-sprint-board-61015de9a849530010b49fc6/issues/celo-org/celo-monorepo/9548)). Precomputing is in this case expensive since we lookup revocation status in storage per attestation found. The alternative for now is: (1) a function for getting all attestations (incl. unrevoked), which precomputes the length of the array and then fills, (2) existing lookup function which takes in the max* param and trims if needed + (3) a separate helper function that does the lookup of # unrevoked attestations. This would allow users who "just want 1" attestation to skip precomputation and just use a max* of 1 (or whatever other small number), but then users who want all unrevoked can choose to do the expensive precomputation and pass in the result to max*, which prevents any trimming/reallocation.

Reply:
I would definitly benchmark this implementation vs the precomputation approach. If precomputing requires additional storage accesses, then it's quite possible that this approach is cheaper. Of course, the UX of it is annoying to specify a length, so the suggestion above of have an option to precompute such that you don't need to select this value is nice.